<section>
<div>
<p class="space-below"><img src="/i/species/pic_Sus_scrofa.png" style="height:75px;width:125px" class="float-left" alt="Pig" title="Oink" />
The Pig gEVAL_10_2 viewer represents the Sscrofa10.2 assembly released in late 2011.  Built using a hybrid de novo assembly involving a clone by clone sequencing strategy suplemented with illumina whole genome shotgun reads, the draft genome comprises of 2.60Gb assigned to chromosomes with a further 212 Mb in unplaced scaffolds.</p> 
</div>
<div><br><br>Further information regarding this assembly can be found in the publication <a href="http://www.nature.com/nature/journal/v491/n7424/full/nature11622.html"><small>Groenen MA et al. Nature 2012</small></a>.</div>
<div><br><p>A live status of the current tiling path can be found using the webtool <a href="http://www.sanger.ac.uk/cgi-bin/humpub/chromoview">Chromoview</a>.</p></div>
</section>

<section>
<div>
<p><a href="http://vega.sanger.ac.uk/Sus_scrofa/">
<img src="/img/vega_small.gif" alt="Vega logo" style="width:83px;height:30px" title="Vega - Vertebrate Genome Annotation database" /></a>
Manual annotations of the pig genome can be found in 
<a href="http://vega.sanger.ac.uk/Sus_scrofa/">Vega</a></p>
</div>
</section>

