<section><div>
<p>The <a href="http://www.sanger.ac.uk/Projects/D_rerio/">zebrafish
genome project</a> is a collaboration between the Sanger Institute and
the zebrafish community, announced during the <a
href="http://www.sanger.ac.uk/about/press/2000/001121.shtml">Sanger Institute
Zebrafish Workshop 2000</a> and was started in February 2001.</p>

<img src="/i/species/pic_Danio_rerio.png" style="width:121px;height:50px"
class="float-left" alt="Danio rerio" title="ploop ploop" />
</div>

<div class="space-below"> 
<br>The zebrafish WGS31 gEVAL viewer represents the whole genome sequence assembly generated from a combinatory approach invovling two double haploid Tubingen fish.  This assembly was created using an Illumina strategy of one double haploid fish (289 Million reads, 30X) coupled with a traditional capillary strategy from a second related double haploid fish (12.2 Million reads, 7.5X).  WGS31 was used in the building of the zebrafish genome assembly Zv9, by integrating wgs sequence into gap regions in the clone assembly.  </div>
<div>
<br><p>For more information regarding the use of WGS31 in Zv9, please view this <a href="http://www.sanger.ac.uk/Projects/D_rerio/Zv9_assembly_information.shtml">page</a></p>
</div>
</section>

<section class="float-left">
<div>
<h4>PGP Viewer Assembly Statistics</h4>
<dt>number of contigs: 119,119</dt>
<dt>number of supercontigs: 32,031</dt>
</div>
</section>

