<section><div>
<p>The <a href="http://www.sanger.ac.uk/Projects/D_rerio/">zebrafish
genome project</a> is a collaboration between the Sanger Institute and
the zebrafish community, announced during the <a
href="http://www.sanger.ac.uk/about/press/2000/001121.shtml">Sanger Institute
Zebrafish Workshop 2000</a> and was started in February 2001.</p>

<img src="/i/species/pic_Danio_rerio.png" style="width:121px;height:50px"
class="float-left" alt="Danio rerio" title="ploop ploop" />
</div>

<div class="space-below"> 
<br>The zebrafish WGS29 gEVAL viewer represents the whole genome sequence assembly generated from a single Tuebingen, double haploid zebrafish.  The assembly is based on 13,756,367 reads comprising 10,891,216,277 bp with a coverage of 5.5x. Phusion and phrap was used to build the consenus contigs, with read pair information used to build supercontigs.  Small supercontigs with less than 3 reads or smaller than 0.5 kb were rejected.  WGS29 was used in the building of the zebrafish genome assembly Zv7, by integrating wgs sequence into gap regions in the clone assembly.  </div>
<div>
<br><p>For more information regarding the use of WGS29 in Zv7, please view this <a href="http://www.sanger.ac.uk/Projects/D_rerio/Zv7_assembly_information.shtml">page</a></p>
</div>
</section>

<section class="float-left">
<div>
<h4>PGP Viewer Assembly Statistics</h4>
<dt>number of contigs: 131,846</dt>
<dt>number of supercontigs: 22,961</dt>
</div>
</section>

