<html>
<head><title>gEVAL Data</title></head>
<body>
<h1>gEVAL Data Access Policies</h1>
<p>
The gEVAL browser uses the EnsEMBL web framework at its core, but modified and hosted at the Wellcome Trust Sanger Institute.  Thus this site follows the WTSI data sharing policy, as outlined <a href="http://www.sanger.ac.uk/about/who-we-are/policies/open-access-science">here</a>.
</p>

<p>
Users should be aware that certain genome assemblies not produced or managed by the Genome Reference Informatics Team or Sanger Institute may have specific policies of their use, unless they have been already widely distributed among the public sequence data repositories.
</p>
<h2>Mouse Genomes Project</h2>
<p>
The Mouse Genomes Project while conducted at the Sanger Institute, is part of a larger consortia and as such have a specific usage policy as outlined below.<br>   

The Mouse Genomes Project releases sequence data, SNPs and other
variant calls as a service to the research community.
</p>
<p>
These data are released in accordance with the Fort Lauderdale
agreement and Toronto agreements. As producers of these data we
reserve the right to be the first to publish a genome-wide analysis
of the data we have generated.
</p>
<p>
The pre-publication data that we release via this website is
embargoed for publication except for analyses of single chromosomes
in single strains or single gene loci across multiple strains. We
strongly encourage researchers to contact us if there are any queries
about referencing or publishing analysis based on pre-publication
data obtained via this website.
</p>

<p>
More information on the Wellcome Trust Sanger Institute's data
sharing policy: <a href="http://www.sanger.ac.uk/datasharing/#guidelines">Guidelines</a>.
</p>
<p>
<strong>NOTE:</strong> These assembled chromosomes are released as unpublished, 
preliminary and incomplete sequences and as such they have not been submitted to 
the accessioned in the public genome sequence repositories (INSDC). The 
assembled sequences will be fully accessioned in public repositories at the time 
of publication of the sequences and are of sufficient quality to be considered 
full draft genome sequences.
</p>

</body>
</head>
